platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/recipe/PermissionsCheckerImpl.java

```diff
@@ -74,4 +79,15 @@ public boolean hasAccess(ManagedRecipe recipe, String userId, String permission)
         }
         return false;
     }
+
+    public void checkPublicPermission(PermissionsDescriptor permissions) throws ForbiddenException {
+        final User user = EnvironmentContext.getCurrent().getUser();
+        if (!user.isMemberOf("system/admin") && !user.isMemberOf("system/manager")) {
+            for (GroupDescriptor group : permissions.getGroups()) {
+                if ("public".equalsIgnoreCase(group.getName()) && group.getAcl().contains("search")) {
```

Comment:
I mean that each code should be simplified as well. It isn't comfortable to read code if it has many nested levels. It's just personal habbit.